dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/index/SearchPermissionsFinder.java

```diff
@@ -0,0 +1,207 @@
+     */
+    public SearchPermissions findFileMetadataPermsFromDatasetVersion(DatasetVersion version) {
+        Dataset dataset = version.getDataset();
+        List<String> perms = new ArrayList<>();
+        
+        Instant publicFrom = SearchPermissions.NEVER_PUBLIC;
+        
+        if (version.isReleased() && !dataset.getEmbargoDate().isDefined()) {
+            publicFrom = SearchPermissions.ALWAYS_PUBLIC;
+        } else if (version.isReleased()) {
+            publicFrom = dataset.getEmbargoDate()
+                    .map(d -> d.toInstant())
+                    .getOrElse(SearchPermissions.NEVER_PUBLIC);
+        }
+        
+        
+        if (publicFrom != SearchPermissions.ALWAYS_PUBLIC) {
+            perms.addAll(findDvObjectPerms(dataset));
+        }
+        
+        return new SearchPermissions(perms, publicFrom);
+    }
+    
+    
+    /**
+     * Returns {@link DatasetVersion}s from the given {@link Dataset}
+     * that needs permission objects indexing.
+     * <p>
+     * Note that returned versions should reflect what is in
+     * {@link IndexServiceBean#indexDataset(Dataset, boolean)}
+     */
+    public Set<DatasetVersion> extractVersionsForPermissionIndexing(Dataset dataset) {
+        Set<DatasetVersion> versionsToIndex = new HashSet<>();
+        
+        DatasetVersion releasedVersion = dataset.getReleasedVersion();
+        DatasetVersion latestVersion = dataset.getLatestVersion();
+        
+        
+        if (releasedVersion != null) {
+            versionsToIndex.add(releasedVersion);
+        }
+        
+        if (latestVersion.getVersionState() == VersionState.DRAFT) {
+            versionsToIndex.add(latestVersion);
+        }
+        if (latestVersion.getVersionState() == VersionState.DEACCESSIONED && releasedVersion == null) {
+            versionsToIndex.add(latestVersion);
+        }
+        
+        return versionsToIndex;
+    }
+
+
+    // -------------------- PRIVATE --------------------
+    
+    private List<String> findDvObjectPerms(DvObject dvObject) {
+        List<String> permStrings = new ArrayList<>();
+        Map<String, RoleAssignee> roleAssigneeCache = new HashMap<>(100);
+        
+        Set<RoleAssignment> roleAssignments = rolesSvc.rolesAssignments(dvObject);
+        for (RoleAssignment roleAssignment : roleAssignments) {
+            logger.fine("role assignment on dvObject " + dvObject.getId() + ": " + roleAssignment.getAssigneeIdentifier());
+            if (roleAssignment.getRole().permissions().contains(getRequiredSearchPermission(dvObject))) {
+                
+                roleAssigneeCache.computeIfAbsent(roleAssignment.getAssigneeIdentifier(), identifier -> roleAssigneeService.getRoleAssignee(identifier));
+                RoleAssignee userOrGroup = roleAssigneeCache.get(roleAssignment.getAssigneeIdentifier());
+                
+                String indexableUserOrGroupPermissionString = getIndexableStringForUserOrGroup(userOrGroup);
```

Comment:
This can be also in `Option.of(getIndexableStringForUserOrGroup(userOrGroup))
                        .peek(permStrings::add);` format.